File: LeetCodeProblemsGo/LeetCode3634MinimumRemovalsToBalanceArray/LeetCode3634MinimumRemovalsToBalanceArray.go
package LeetCode3634MinimumRemovalsToBalanceArray

import (
	"fmt"
	"slices"
)

func minRemoval(nums []int, k int) int {
	slices.Sort(nums)
	length := len(nums)
	maxLength := 0
	currentLength := 0
	begin, end, mid, found := 0, 0, 0, 0
	for i := 0; i < length; i++ {
		begin = i
		end = length - 1
		found = -1
		for begin <= end {
			mid = (begin + end) >> 1
			if nums[mid] <= k*nums[i] {
				found = max(found, mid)
				begin = mid + 1
				continue
			}
			end = mid - 1

		}
		currentLength = found - i + 1
		maxLength = max(currentLength, maxLength)
	}
	return length - maxLength
}
func Calculate() {
	nums := []int{1, 34, 23}
	k := 2
	fmt.Println(minRemoval(nums, k))
}

File: LeetCodeProblemsCMake/Problems/LeetCode3634MinimumRemovalsToBalanceArray.cpp
//
// Created by saeed on 2/6/26.
//

#include "LeetCode3634MinimumRemovalsToBalanceArray.h"

#include <algorithm>
using namespace std;

int LeetCode3634MinimumRemovalsToBalanceArray::minRemoval(vector<int> &nums, int k) {
    ranges::sort(nums);
    int secondPointer{0};
    int currentLength{0};
    long maxAllowed{0};
    int maxLength{1};
    int length = static_cast<int>(nums.size());
    for (int i = 0; i < length - 1; ++i) {
        maxAllowed = k;
        maxAllowed*= nums[i];
        while (secondPointer < length - 1 && nums[secondPointer] <= maxAllowed) {
            secondPointer++;
        }
        currentLength = secondPointer - i + 1;
        if (nums[secondPointer] > maxAllowed) {
            --currentLength;
        }
        maxLength = max(maxLength, currentLength);
    }
    return length - maxLength;
}
